src/extra/components/las.py

```diff
@@ -0,0 +1,490 @@
+                  f'be correlated to BAM corrections. Please use pulse '
+                  f'informations including FEL pulses, such as those provided '
+                  f'by the XrayPulses or PumpProbePulses components.',
+                  file=sys.stderr)
+            self._bam_src = self._bam_delay_key = None  # Disable BAM.
+
+        self._pulses = pulses
+        self._trigger_ref = ref_trigger
+        self._stage_ref = ref_stage
+        self._bam_ref = ref_bam
+        self._delay_ref = ref_delay
+
+    @classmethod
+    def _try_delay_source(cls, data, source, defaults, name):
+        if source is False:
+            # Explicitly disabled.
+            return False
+
+        if source is None:
+            # Auto-detection.
+            if (default := defaults.get(name)) is not None:
+                # Existing default.
+                source = default
+            else:
+                # No default, disable.
+                return False
+
+        if source not in data.all_sources:
+            # Source missing.
+            raise ValueError(f'expected source {source} not found in data, '
+                             f'please explicit {name} source or disable with '
+                             '{name}=False')
+
+        return source
+
+    def _get_train_delays(self, labelled, by_pulse, pulse_dim, kd, ref, scale):
+        delays = (kd.ndarray() - ref) * scale
+
+        if by_pulse:
+            if self._pulses is None:
+                raise ValueError('component is not initialized with pulse '
+                                 'information')
+
+            delays = delays.repeat(self._pulses.pulse_counts())
+
+            if labelled:
+                # If labelled, build a series now from the ndarray.
+                delays = pd.Series(
+                    delays, index=self._pulses.build_pulse_index(pulse_dim))
+
+        elif labelled:
+            # Labelled series, but not by train.
+            delays = pd.Series(delays, index=kd.train_id_coordinates())
+
+        return delays
+
+    def __repr__(self):
+        use_labels = []
+
+        if self._trigger_src is not None:
+            use_labels.append(f'trigger={self._trigger_src.source}'
+                              f'[ref={self._trigger_ref}]')
+
+        if self._stage_src is not None:
+            use_labels.append(f'stage={self._stage_src.source}'
+                              f'[ref={self._stage_ref}]')
+
+        if self._bam_src is not None:
+            use_labels.append(f'bam={self._bam_src.source}'
+                              f'[ref={self._bam_ref}]')
+
+        if self._pulses is not None:
+            use_labels.append(f'pulses={repr(self._pulses)}')
+
+        return '<{} using\n  {}>'.format(
+            type(self).__name__, '\n  '.join(use_labels))
+
+    @property
+    def stage_to_time(self) -> float:
+        """millimeter to seconds"""
+        return -2e-3 / speed_of_light
```

Comment:
Exactly, the -2 comes from the fact the motors moves the arm linearly, while the light takes that distance twice. I integrated it there as this is the case for both MID's and SQS' stage so far.

I suspect FXE is then actually calling their *trigger delay* an *optical delay line* and we're mixing up names here. `DoocsOpticalDelay` is the device class used for trigger delay (see comment in initializer in the `trigger` section) in the respective `LAn` topics in units of picoseconds. Likely this is enough for FXE's needs, and either there is another *stage* they're not using often or not at all.

It tells us though that we will face a lot of naming issues here. I tried to stick to naming as the LAS group uses, but no doubt there will be other terms used locally at instruments.

Reply:
Summary from a brief in-person discussion on this:

* Likely we'll run into more ambiguities regarding names of individual delay sources
* `DoocsOpticalDelay` seems to actually include beam paths in the case of FXE, so not be a trigger delay like for SQS
* We can merge as is to make it work for SQS (and potentially MID) and laster revisit when we have data for all instruments - likely we would make the labels flexible rather than fixed and not change the fundamental logic

Reply:
We agreed offline to make the device less configurable for now, so it will only work for the instruments with devices already defined, until we can work out what exactly we need for other instruments, so as to avoid publishing an API which we soon have to redesign.